### ISSUE TITLE
Warning when creating `foo` and `foo:includes` collections

### ISSUE THREAD

**Issue by johnhaddon** [PixarAnimationStudios/USD#3251]
As discussed in https://forum.aousd.org/t/guidance-for-api-schema-instance-names/1498, USD issues a warning if both `foo` and `foo:includes` collections are authored on the same prim. This can be seen by executing the following example code :

```
stage = Usd.Stage.CreateInMemory( "test" )
prim = stage.DefinePrim( "/Test" )
Usd.CollectionAPI.Apply( prim, "foo" )
Usd.CollectionAPI.Apply( prim, "foo:includes" )
Usd.CollectionAPI.GetAllCollections( prim )
```

Which yields the following warning :

```
Warning: in _PropertyTypesMatch at line 257 of /.../primDefinition.cpp -- Cannot compose schema specs: Schema relationship spec at path '/CollectionAPI.collection:__INSTANCE_NAME__:includes' in layer 'anon:0x7f4f19446080' is a different spec type than schema attribute spec at path '/CollectionAPI.collection:__INSTANCE_NAME__' in layer 'anon:0x7f4f19446080'.
```

It would be great if this restriction could be removed.

**Comment by jesschimein** [comment]
Filed as internal issue #USD-10033

**Comment by johnhaddon** [comment]
Just a quick note to say that I'm still seeing this in USD 26.03. Is there any hope of a fix?

**Comment by spiffmon** [comment]
Hi @johnhaddon , the short answer is that there is no possible "fix" to the warning you are seeing - you are creating scene description that is not robustly evaluatable.  Some details:

1. The issue reported by @nvmkuruc in the referenced forum post can and should be addressed
2. However, there is a very general problem with *combinations* of API schemas on a prim, regardless of whether they are single-apply or multiple-apply: any property in USD can only have a single type; therefore if two different schemas name the same property, but assign it **two different types** (or, as in your case, one is a relationship and the other is an attribute), there is no correct way to interpret the data for **both schemas** on the same prim.  OpenUSD follows the "strongest wins" pattern in that the prim definition we construct will contain the property from the "stronger applied" (or typed) schema, but when you view the prim's data through the weaker applied schema object, you'll have problems, as will downstream consumers.  Our current response is to warn you about it at runtime, though we think an additional thing we want to do is add a validation error for each such occurrence.
3. When, in the forum-thread, I said that we were being "overly restrictive", I was referring to the logic in the schema classes' `CanApply()` methods, which we hope to make more targetted, and follow the same logic that the validation error would.  But I guess I'd also suggest that you use `CanApply()` even in its current form, which would tell you (I believe) you should not create both a `foo` and `foo:includes` collection (it may forbid the latter altogether currently, though as you point it, it is not actually problematic in and of itself).

Does that make sense?

**Comment by johnhaddon** [comment]
Thanks @spiffmon, that does make sense. I had assumed that the clash in opinions on the type of the property was the problem, but had got the wrong end of the stick from your forum reply, thinking that you were saying it was fixable. Otherwise I wouldn't have opened the issue - I'll close now.

If I were to quibble though, I'd have to say we're not "totally clear to use colons in instance names". It seems we're clear to use them as long as we don't accidentally cause a collision, which requires being aware of all the properties the schema might make. In our case, the user is providing the collection names indirectly (in Gaffer), and we just have to hope they don't inadvertently cause a collision. I think it's unlikely that they will, but it's not something we really have control over.

**Comment by spiffmon** [comment]
No, you **are** clear to use colons, @johnhaddon - the underlying problem actually has nothing to do with property namespaces or colons (and to your point about "property awareness", the consistent use of property namespaces in schemas is one of the best ways to **avoid** the problem.)  CollectionAPI is a specifically (and uniquely, thus far, in our core schemas) difficult case because it is multi-apply **and** creates a property that consists of just the `schema:instance` name.  But you can get into trouble without colons and with only single-apply schemas:

If `FooAPI` defines a property
```
    token uses = "none"
```

and `BarAPI` defines a property
```
    rel uses 
```

If you apply **both** to the same prim, you will get the exact same warning about `uses` when we try to create a PrimDefinition for the prim.

I wish I could tell you that the pattern we use in CollectionAPI won't propagate, but we learned several years ago in Presto that it is highly useful to facilitate "bundled relationships" (and connections), where we can, e.g. connect all the attributes in a namespace `inputs:` to same-named attributes on another prim simply by connecting the schema's `opaque inputs` attribute to a single target on another prim (I'm fudging/glossing over a few details - we'll disclose more about this at the right time in OpenExec rollout).

I sympathize with gaffer's position... without a live underlying UsdStage you can't use the schema class's `CanApply()` validator... since CollectionAPI is the _most likely_ place where these kinds of collisions might come up (and for sure, calling your collection itself "includes" or "excludes" is a natural thing to want to do), maybe you can use a hardcoded validator that checks the specific "combination" situation you're reporting?

**Comment by johnhaddon** [comment]
Thanks for the clarifications @spiffmon. `CanApply()` was the missing part of the puzzle for me - I hadn't realised it validated against the property names for the schema. Are there any recommendations for UX if a user does try to make something like `foo:includes`? Should the failure just be communicated, or should some attempt to be made to adjust the name to something valid?
